---
title: "Tursiops RADseq summary"
author: Reid Brennan
output:
  html_document:
    theme: united
    toc: true
    toc_float: true
    center: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(knitr)
library(kableExtra)
```

337 samples and 7577 SNPs from across the northern gulf and western Atlantic. 4356 SNPs after LD thinning

![](../figures/map.png){width=70%}

---

# population structure

First, just run a basic PCA:

![](../figures/pca_popIds_region.png){width=70%}


![](../figures/pca_depth.png){width=70%}

![](../figures/pca_distance.png){width=70%}

We get clustering by shallow and deep/ nearshore and offshore, rather than Atlantic vs. Gulf. Suggesting that nearshore in the Atlantic and Gulf are more similar that geographic promixmate animals nearshore and offshore.

---

The patterns above maybe look like an inversion or something, but the PCA loadings don't indicate this:
![](../figures/manhattan_plot_pc1.png){width=90%}

![](../figures/manhattan_plot_pc2.png){width=90%}


---

## admixture

Sorting the plot below by Atlantic on the left, Gulf on the right. Then shallow to deep by collection location 

Note that the left group in the Atlantic are the nearshore individuals, the bottom left in the PCAs above. 


![](../figures/Admixture_plot.png)

### ordered shallow to deep, following subset for unequal sample sizes:

![](../figures/Admixture_plot_subset.png)

<br>

## Population structure with admixture populations

![](../figures/pca_admixturePops.png){width=70%}


### NJ tree of pairwise genetic distances 

![](../figures/nj_tree.png){width=70%}

<br>
<br>


#### Population assignment conclusion

Both Admixture and DAPC (not shown) indicate 4 populations as the most likely. They also agree with their individual assignments for all but ~5 individuals, who are all fairly admixed. 

<br><br>

## Isolation by distance:

Next, testing if there is isolation by distance at various scales. 

Across all individuals, there is an IBD signal. But I think this is likely driven by underlying comparisons. 

![](../figures/ibd_allpops.png){width=70%}

<br>

Split this out into each individual population across both regions. Within a population we generally see IBD, except for Atlantic Offshore. Inter-population comparisons have no signal. 

![](../figures/ibd_withinpops.png){width=70%}

Tt312 and 13Tt073 are divergent and causing that weird pattern in offshore atlantic

<br>

Can also split into just the Atlantic of Gulf:

![](../figures/ibd_withinregion.png){width=70%}

And split into Coastal, Intermediate, and Offshore:

![](../figures/ibd_withindepth.png){width=70%}

![](../figures/ibd_withindepth_interOnly.png){width=70%}

This shows that the IBD pattern when all individuals are compared is driven primarily by within population comparisons and IDB is not really, or perhaps only weakly, present between the gulf and the atlantic.

---

# Introgression and hybrids

The Intermediate individuals are possibly hybrids, given the results above they're intermediate in genetic distance for nearly all analyses. Here, I'll test this using:

1. f3 statistics
2. D-/f4 statistics
2. treemix
3. triangle plots
4. new hybrids


### f3 statistics:

The f3-statistic explicitly tests whether a taxon of interest C results from admixture between two others (A and B): A significantly negative f3-statistic supports the admixture hypothesis, while a positive value is not informative. In our case, our taxon of interest C is `Intermediate` while A and B are `Coastal` and `Offshore`. 

First, I calculated these statistics with the four populations assignments:


```{r four-pop-table}
# First table data

table1 <- read.table("../analysis/pop_structure/admixtools/f3_stats_fourpop.txt", header=T)

kable(table1, 
      format = "html", 
      caption = "F3-Statistics with Four Populations",
      col.names = c("pop1", "pop2", "pop3", "est", 
                    "se", "z", "p"),
      align = c('c', 'c', 'c', 'c', 'c', 'c', 'c')) %>%
  kable_styling(bootstrap_options = c("striped", "hover"), 
                full_width = FALSE)
```


Remember, positive values are not informative, negative values indicate a population resulting from admixture. No significance here. 

Next, I split into 6 populations:


```{r six-pop-table}
# Second table data
table2 <- read.table("../analysis/pop_structure/admixtools/f3_stats_sixpop.txt", header=T)


# Create and display the table directly
kable(table2, 
      format = "html", 
      caption = "F3-Statistics with Six Populations",
      col.names = c("pop1", "pop2", "pop3", "est", 
                    "se", "z", "p"),
      align = c('c', 'c', 'c', 'c', 'c', 'c', 'c')) %>%
  kable_styling(bootstrap_options = c("striped", "hover"), 
                full_width = FALSE) %>%
  row_spec(c(7, 8), background = "salmon") %>%  # Highlight rows with negative F3
  pack_rows("Intermediate Atlantic", 1, 4) %>%
  pack_rows("Intermediate Gulf", 5, 8)

```


This indicates that the intermediate Gulf population is a result of admixture between the coastal gulf and both the Offshore Atlantic and Gulf populations. There is no evidence in the Atlantic Intermediate population. 

<br>
<br>

### D-/f4 statistics:

D-statistics, or ABBA-BABA tests, test for introgression by looking for violations of incomplete lineage sorting. In short, if we have a tree with an ancestral "A" allele and derived "B" allele in the tree (((P1,P2),P3),O) where O is the outgroup, we should see an "ABBA" or "BABA" pattern at equal frequencies when there is incomplete lineage sorting and no gene flow. If there is an over representation of either ABBA or BABA, this suggests gene flow (see figure below, from the Dsuite tutorial). 
<br>

![Example of ABBA BABA test](../figures/DstatIllustration2.png){width=70%}
<br>
<br>
I ran this test with [Dsuite](https://github.com/millanek/Dsuite), with Aduncus as the outgroup. For the output below, P1 and P2 will always be arranged so that D is positive and indicates geneflow between P2 and P3. P1 and P2 could be flipped which would just flip the sign of D to negative and indicate gene flow between P1 and P3. 


```{r four-pop-dstats}
# Second table data
table2 <- read.table("../analysis/pop_structure/dstats/dstats_fourpop.txt", header=T)

kable(table2, 
      format = "html", 
      caption = "D-Statistics with four Populations",
      #col.names = c("pop1", "pop2", "pop3", "est", 
      #              "se", "z", "p"),
      align = "c") %>%
  kable_styling(bootstrap_options = c("striped", "hover"), 
                full_width = FALSE) %>%
  row_spec(which(table2$p.value_multTesting < 0.05), background = "salmon") #%>%  # Highlight rows with negative F3
  #pack_rows("Intermediate Atlantic", 1, 4) %>%
  #pack_rows("Intermediate Gulf", 5, 8)

```

- There is the strongest evidence for gene flow between the coastal gulf and intermediate population (line 1). 
- There is also gene flow between Coastal Gulf and Offshore
<br>


```{r six-pop-dstats}
# Second table data
table2 <- read.table("../analysis/pop_structure/dstats/dstats_sixpop.txt", header=T)

kable(table2, 
      format = "html", 
      caption = "D-Statistics with six Populations",
      #col.names = c("pop1", "pop2", "pop3", "est", 
      #              "se", "z", "p"),
      align = "c") %>%
  kable_styling(bootstrap_options = c("striped", "hover"), 
                full_width = FALSE) %>%
  row_spec(which(table2$p.value_multTesting < 0.05), background = "salmon") #%>%  # Highlight rows with negative F3
  #pack_rows("Intermediate Atlantic", 1, 4) %>%
  #pack_rows("Intermediate Gulf", 5, 8)

```

- There is geneflow between both Intermediate Atlantic and Intermediate Gulf with Coastal Gulf. 
- Gene flow between Coastal gulf and both offshore populations
- No signal for intermediate and offshore populations


<br><br>



### Treemix

Maximum likelihood tree estimating drift among populations. Migration edges are fit to the tree to improve populations that are a poor fit to the model. Migration gets addes stepwise. You can estimate the number of migration events that improves the model fit best, similar to structure evanno type approaches. 

#### no migration

First fit the trees with 0 migration events:

![Four populations, no migration events](../figures/TreeMix_noMigration_fourpop.png){width=50%}![Six populations, no migration events](../figures/TreeMix_noMigration_sixpop.png){width=50%}

#### Adding migration events:

Best number of migrations events:

- Four populations: 2 migrations events
- Six populations: 4 migration events

The four population result is consistent and clear. There are two migration edges, between intermediate and offshore and the node of intermediate/Coastal Gulf and offshore. This tree is well supported and consistent across runs (100 runs, nearly all show this exact tree, below). 


![](../figures/TreeMix_output_fourpop.png){width=50%}

In contrast, with 6 populations, things are much more uncertain/unstable. For the most likely tree (top left, in figure below) there are migration edges between the intermediate pops and the branch leading to coastal populations. There is also migration from the node of coastal Gulf/intermediate with both offshore populations. The next 5 most likely trees show similar variations on these migration events. Note that the position of the coastal and intermediate populations are unstable across runs. This maybe isn't shocking given that these populations aren't well supported in the other analyses. 

Results that are consistent:

- Offshore Gulf and Offshore Atl are always sister
- There is no migration with the coastal Atlantic population. 
- there is ample migration between Intermediate and offshore populations. 

![](../figures/TreeMix_output_sixpop_top6_LL.png){width=100%}

### Triangle plots

The basic idea behind these is that we can idenfity early generation hybrids by both their ancestry and (interclass) heterozygosity. We consider highly divergent differences (> 0.7 frequency; 0.8 snf 0.9 give similar results) between parental populations Ancestry Informative Markers (AIMs). Then an F1 hybrid would have a hybrid index based on these AIMs of 0.5 (50% of alleles from either parent population). We calculate how many of AIMs in the putative hybrids are heterozygous for ancestry from either parent. For an F1, all loci would be heterozygous, so this value would be 1. With F2, this heterozygosity would drop to ~0.5. and continue to drop if there is backcrossing. 

Here's a nice paper that shows expectations for different scenarios. In short, if we follow the expectation curve in the plot below, it is likely due to admixture and not isolation by distance or a similar process. In contrast, there should be no relationship between hybrid index and heterozygosity when admixture has not occurred and IBD is the main feature of the data  https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.14039




![Four population assignment](../figures/triangle_plot_fourpop.png){width=100%}

<br><br>

![Six population assignment](../figures/triangle_plot_sixpop.png){width=100%}

<br><br>
There are no F1 hybrids in these data, but the rest of the variation is likely due to admixture, not neutral IDB. 

<br><br><br><br><br><br>

# still to do

1. Genomic cline analysis
  - using introgress, bgc, or similar (from Gompert). 
2. gene environment associations
3. general selection scans within species. 


