---
title: "Tursiops RADseq summary"
author: Reid Brennan
date: "`r Sys.Date()`"
output:
  html_document:
    theme: united
    toc: true
    toc_float: true
    center: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(knitr)
library(kableExtra)
```


Results of initial analysis for Tursiops RADseq analysis from the Gulf and Atlantic. Overall goals are to understand population structure, look for evidence of hybridization, and determine range extent of the species. 

- Initial number of samples: 375
- Samples dropped due to sequencing quality/coverage: 3 + 24
- Samples dropped because they're strandings: 11
- 337 samples in the data
- Number of SNPs after filtering:7577
- Number of SNPs after LD thinning:4356

[Link to scripts on github](https://github.com/rsbrennan-NOAA/SEFSC-MMTD-Tursiops_RAD_popgen)

See map of sample locations below. 


![Map of sample locations](../figures/map.png){width=70%}

---

# Population structure

## PCA

First, just run a basic PCA with all samples, colored by region:

![](../figures/pca_popIds_region.png){width=70%}


<br>

### PCA colored by depth

![](../figures/pca_depth.png){width=70%}

## PCA colored by distance from shore

![](../figures/pca_distance.png){width=70%}

We get clustering by shallow and deep/ nearshore and offshore, rather than Atlantic vs. Gulf. Suggesting that nearshore in the Atlantic and Gulf are more similar that geographic promixmate animals nearshore and offshore.

---

### Checking for an inversion

The patterns above maybe look like an inversion. If we saw this, we'd see the PCA loadings clustered in a single location in the genome, but the PCA loadings don't indicate this, instead we see the loadings distributed across the genome.

![](../figures/manhattan_plot_pc1.png){width=90%}

![](../figures/manhattan_plot_pc2.png){width=90%}


---

<br><br>

## Admixture

Running multiple K's with the cross-validation error for each. Lowest error indicates the most likely K. It is basically withholding a subset of the genotypes then predicting their values and compares this to the withheld data. 

Sorting the plot below by Atlantic on the left, Gulf on the right. Then shallow to deep by collection location 

Note that the left group in the Atlantic are the nearshore individuals, the bottom left in the PCAs above. 


![](../figures/Admixture_plot.png)

### ordered shallow to deep, following subset for unequal sample sizes:

![](../figures/Admixture_plot_subset.png)

<br>

## PCA with populations from Admixture and DAPC

![](../figures/pca_admixturePops.png){width=70%} ![](../figures/pca_DAPCPops.png){width=70%} 




### NJ tree of pairwise genetic distances 

![](../figures/nj_tree.png){width=70%}

<br>

There is a group of offshore individuals, particularly in the gulf, that look like they might be hybrids, based on the PCA and nj tree. 

<br>
<br>

### Map of population assignments


![](../figures/map_allpops.png){width=100%}

![](../figures/map_atlantic.png){width=100%}

![](../figures/map_gulf.png){width=100%}


## Population assignment conclusion

- Both Admixture and DAPC (not shown) indicate 4 populations as the most likely. They also agree with their individual assignments for all but ~5 individuals, who are all fairly admixed. 
- The four popoulations are:
  - Coastal Gulf
  - Coastal Atlantic
  - Intermediate
  - Offshore

For the rest of the analyses, I'll run both the four population analysis as well as a six population analysis, splitting the intermediate and offshore into Gulf and Atlantic. I think this makes sense biologically and is justified given our hypotheses going into the analysis. 


<br><br>

## Isolation by distance:

Next, testing if there is isolation by distance at various scales. For all I'm using PCA-based genetic distance with 64 PCs (based on [Shirk et al.](https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12684)), but these values are very similar to Plink and Euclidian values. 

### IDB across all individuals

Across all individuals, there is an IBD signal. But I think this is likely driven by underlying comparisons. 

![](../figures/ibd_allpops.png){width=70%}

<br>

### IBD: Four population analysis

![](../figures/ibd_fourpops.png){width=70%}

### IBD: Six population analysis

Split this out into each individual population across both regions. Within a population we generally see IBD, except for Atlantic Offshore. Inter-population comparisons have no signal. 

![](../figures/ibd_sixpops.png){width=70%}


<br>

### IBD: Atlantic and Gulf

![](../figures/ibd_withinregion.png){width=70%}

### IBD: Combining Atlantic and Gulf

![](../figures/ibd_withindepth.png){width=70%}

### IBD: Comparing only between populations

![](../figures/ibd_withindepth_interOnly.png){width=70%}

<br>

### IBD conclusion

- strong signal of IBD within both the four and six populations. 
- Between the Gulf and Atlantic, there is relatively minimal IBD. 


---

<br>
<br>

## Pairwise Fst

Above diagonal, Fst, below, p-values from 500 bootstrap permutations. 

![](../figures/fst.png){width=70%}


![](../figures/fst_fourpop.png){width=50%}

<br><br>

# Introgression and hybrids

The Intermediate individuals are possibly hybrids, given the results above they're intermediate in genetic distance for nearly all analyses. Here, I'll test this more explicitly using:

1. f3 statistics
2. D statistics
2. treemix
3. triangle plots
4. new hybrids


### f3 statistics:

The f3-statistic explicitly tests whether a taxon of interest results from admixture between two others: A significantly negative f3-statistic supports the admixture hypothesis, while a positive value is not informative. In our case, our taxon of interest (pop1) is `Intermediate` while pop1 and pop2 are `Coastal` and `Offshore`. 

First, I calculated these statistics with the four populations assignments:


```{r four-pop-table}
# First table data

table1 <- read.table("../analysis/pop_structure/admixtools/f3_stats_fourpop.txt", header=T)

kable(table1, 
      format = "html", 
      caption = "F3-Statistics with Four Populations",
      col.names = c("pop1", "pop2", "pop3", "est", 
                    "se", "z", "p"),
      align = c('c', 'c', 'c', 'c', 'c', 'c', 'c')) %>%
      row_spec(which(table1$p < 0.05 & table1$est < 0), background = "#f8d0d8") %>%  # Highlight rows with negative F3

  kable_styling(bootstrap_options = c("striped", "hover"), 
                full_width = FALSE)
```


Remember, positive values are not informative, negative values indicate a population resulting from admixture.

Next, I split into 6 populations:


```{r six-pop-table}
# Second table data
table2 <- read.table("../analysis/pop_structure/admixtools/f3_stats_sixpop.txt", header=T)


# Create and display the table directly
kable(table2, 
      format = "html", 
      caption = "F3-Statistics with Six Populations",
      col.names = c("pop1", "pop2", "pop3", "est", 
                    "se", "z", "p"),
      align = c('c', 'c', 'c', 'c', 'c', 'c', 'c')) %>%
  kable_styling(bootstrap_options = c("striped", "hover"), 
                full_width = FALSE) %>%
  #row_spec(c(7, 8), background = "#f8d0d8") %>%  # Highlight rows with negative F3
    row_spec(which(table2$p < 0.05 & table2$est < 0), background = "#f8d0d8") %>%  # Highlight rows with negative F3
  pack_rows("Intermediate Atlantic", 1, 4) %>%
  pack_rows("Intermediate Gulf", 5, 8)

```


This indicates that the intermediate Gulf population is a result of admixture between the coastal gulf and both the Offshore Atlantic and Gulf populations. There is no evidence in the Atlantic Intermediate population. 

<br>
<br>

### D-/f4 statistics:

D-statistics, or ABBA-BABA tests, test for introgression by looking for deviations from incomplete lineage sorting. In short, if we have a tree with an ancestral "A" allele and derived "B" allele in the tree (((P1,P2),P3),O) where O is the outgroup, we should see an "ABBA" or "BABA" pattern at equal frequencies when there is incomplete lineage sorting and no gene flow. If there is an over representation of either ABBA or BABA, this suggests gene flow (see figure below, from the Dsuite tutorial). 
<br>

![Example of ABBA BABA test](../figures/DstatIllustration2.png){width=70%}
<br>
<br>
I ran this test with [Dsuite](https://github.com/millanek/Dsuite), with Aduncus as the outgroup. For the output below, P1 and P2 will always be arranged so that D is positive and indicates geneflow between P2 and P3. P1 and P2 could be flipped which would just flip the sign of D to negative and indicate gene flow between P1 and P3. 


```{r four-pop-dstats}
# Second table data
table2 <- read.table("../analysis/pop_structure/dstats/dstats_fourpop.txt", header=T)

kable(table2, 
      format = "html", 
      caption = "D-Statistics with four Populations",
      #col.names = c("pop1", "pop2", "pop3", "est", 
      #              "se", "z", "p"),
      align = "c") %>%
  kable_styling(bootstrap_options = c("striped", "hover"), 
                full_width = FALSE) %>%
  row_spec(which(table2$p.value_multTesting < 0.05), background = "#f8d0d8") #%>%  # Highlight rows with negative F3
  #pack_rows("Intermediate Atlantic", 1, 4) %>%
  #pack_rows("Intermediate Gulf", 5, 8)

```

- There is the strongest evidence for gene flow between the coastal gulf and intermediate population (line 1). 
- There is also gene flow between Coastal Gulf and Offshore
<br>


```{r six-pop-dstats}
# Second table data
table2 <- read.table("../analysis/pop_structure/dstats/dstats_sixpop.txt", header=T)

kable(table2, 
      format = "html", 
      caption = "D-Statistics with six Populations",
      #col.names = c("pop1", "pop2", "pop3", "est", 
      #              "se", "z", "p"),
      align = "c") %>%
  kable_styling(bootstrap_options = c("striped", "hover"), 
                full_width = FALSE) %>%
  row_spec(which(table2$p.value_multTesting < 0.05), background = "#f8d0d8") #%>%  # Highlight rows with negative F3
  #pack_rows("Intermediate Atlantic", 1, 4) %>%
  #pack_rows("Intermediate Gulf", 5, 8)

```

- There is geneflow between both Intermediate Atlantic and Intermediate Gulf with Coastal Gulf. 
- Gene flow between Coastal gulf and both offshore populations
- No signal for intermediate and offshore populations


<br><br>



### Treemix

Maximum likelihood tree estimating drift among populations. Migration edges are fit to the tree to improve populations that are a poor fit to the model. Migration gets addes stepwise. You can estimate the number of migration events that improves the model fit best, similar to structure evanno type approaches. 

#### no migration

First fit the trees with 0 migration events:

![Four populations, no migration events](../figures/TreeMix_noMigration_fourpop.png){width=50%}![Six populations, no migration events](../figures/TreeMix_noMigration_sixpop.png){width=50%}

#### Adding migration events:

Best number of migrations events:

- Four populations: 2 migrations events
- Six populations: 4 migration events

The four population result is consistent and clear. There are two migration edges, between intermediate and offshore and the node of intermediate/Coastal Gulf and offshore. This tree is well supported and consistent across runs (100 runs, nearly all show this exact tree, below). 


![](../figures/TreeMix_output_fourpop.png){width=50%}

In contrast, with 6 populations, things are much more uncertain/unstable. For the most likely tree (top left, in figure below) there are migration edges between the intermediate pops and the branch leading to coastal populations. There is also migration from the node of coastal Gulf/intermediate with both offshore populations. The next 5 most likely trees show similar variations on these migration events. Note that the position of the coastal and intermediate populations are unstable across runs. This maybe isn't shocking given that these populations aren't well supported in the other analyses. 

Results that are consistent:

- Offshore Gulf and Offshore Atl are always sister
- There is no migration with the coastal Atlantic population. 
- there is ample migration between Intermediate and offshore populations. 

![](../figures/TreeMix_output_sixpop_top6_LL.png){width=100%}

### Triangle plots

The basic idea behind these is that we can idenfity early generation hybrids by both their ancestry and (interclass) heterozygosity. We consider highly divergent differences (> 0.7 frequency; 0.8 snf 0.9 give similar results) between parental populations Ancestry Informative Markers (AIMs). Then an F1 hybrid would have a hybrid index based on these AIMs of 0.5 (50% of alleles from either parent population). We calculate how many of AIMs in the putative hybrids are heterozygous for ancestry from either parent. For an F1, all loci would be heterozygous, so this value would be 1. With F2, this heterozygosity would drop to ~0.5. and continue to drop if there is backcrossing. 

Here's a nice paper that shows expectations for different scenarios. In short, if we follow the expectation curve in the plot below, it is likely due to admixture and not isolation by distance or a similar process. In contrast, there should be no relationship between hybrid index and heterozygosity when admixture has not occurred and IBD is the main feature of the data  https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.14039




![Four population assignment](../figures/triangle_plot_fourpop.png){width=100%}

<br><br>

![Six population assignment](../figures/triangle_plot_sixpop.png){width=100%}

#### with putative hybrids from PCA

The "hybrids" below are those from the PCA that appear to be a separate cluster in the PCA for the offshore gulf. Note they do look to be hybrids of some sort. 

![Six population assignment](../figures/triangle_plot_sixpop_hybrids.png){width=100%}


<br><br>
There are no F1 hybrids in these data, but the rest of the variation is likely due to admixture, not neutral IDB. There could be some recent-ish hybrids, but a little unclear from these results. Maybe new hybrids could help clarify. 

<br><br><br>

### NewHybrids

For first pass, just using 90 loci that are divergent between offshore Gulf and Coastal Gulf. I ran 10 independent runs of NewHybrids. The median values of these runs are plotted below, where `High_Confidence` are individuals with a posterior > 0.99, and `Low_Confidence` are individuals with a posterior below 0.99. Note that all individuals were consistently assigned to the same group across all runs. The count of individuals in each bin are below the boxplot.

![Six population assignment](../figures/newhybrids.png){width=100%}


<br>

And adding the assignments to the triangle plot:

![Six population assignment](../figures/triangle_plot_sixpop_Newhybrids.png){width=100%}

Again, there are no F1 hybrids, but there are some recent hybrids. 

<br> 

#### Mapping putative hybrids as red circles

The NewHybrids assignments are shown as text in the circle. 

- F2 = F2
- B1 = Backcross 1
- B2 = Backcross 2
- No Label = no New Hybrids Assignment

![](../figures/map_allpops_hybrids.png){width=100%}


<br><br>

# still to do

1. new hybrids
2. Population structure within species
3. Genomic cline analysis
  - using introgress, bgc, or similar (i.e., Gompert). 
4. gene environment associations
5. general selection scans within species. 


# conclusions

as of 2025-03-26

- Erebennus range includes both the Atlantic and the Gulf. This is a single species. 
- Truncatus range is both the Atlantic and the Gulf. This is a single species.
- There is substantial and significant divergence between the Gulf and the Atlantic for both of these species. There is more divergence between Erebennus in the Gulf vs Atantic than for Truncatus, which makes sense. 
- The intermediate population is a result of hybridization between Erebennus and Truncatus.
  - There is minimal structure between the Gulf and the Atlantic.  
  - There has likely been lots of backcrossing between the Erebennus and the Intermediate populations. Particularly in the Gulf. 
  - The intermediate populations are likely Erebennus, potentially a subspecies. 

# outstanding questions for discussion

- Stranded individuals were excluded. Should we reconsider this? There are only 11. 
- Do we agree that there is sufficient evidence that erebennus extends to the Gulf?
- What should we consider the intermediate population. 
- Additional analyses to consider? 
- Changes to existing analyses?
